### ISSUE TITLE
Doped Clifford Sampling - Clifford Doping Validation

### ISSUE THREAD

**Issue by jchung2020** [quantum-advantage-tracker/quantum-advantage-tracker.github.io#228]
### Name

ibm_boston_dcs_nq70_depth70_checks27

### Circuit

doped_random_graph_sampling_nq70_depth70_checks27

### Value

0.32

### Method

Quantum

### Method proof

We provide a summary of the doped clifford sampling problem below, a more detailed writeup is available in the [circuit instance description](https://github.com/quantum-advantage-tracker/quantum-advantage-tracker.github.io/tree/main/data/classically-verifiable-problems/circuit-models/doped_random_graph_sampling).

## Doped Clifford Sampling

We present an experiment to demonstrate quantum advantage by sampling from T-doped random graph states in the Z-basis. We prove that strong simulation - computing amplitudes of measurement outcomes - on this circuit family is average and worst case #P-hard (to appear). This is an extension on previous work demonstrating #P-hardness for random *regular* graphs measured in a random product basis<sup>1</sup>. While this bears similarities to random circuit sampling<sup>2</sup>, by preparing T-doped random graph states, we can:

1. Generate highly entangled states.
2. Employ spacetime error detection<sup>3</sup> to post-select errored shots and achieve high fidelities, even at large circuit widths and depths
3. Verify fidelity efficiently with direct fidelity estimation<sup>4</sup>, which requires fewer assumptions about the noise and does not extrapolate from smaller circuits, as for cross-entropy benchmarking (XEB)
4. Tune in and out of the classically verifiable regime by controlling the number of T gates used during the experiment

While we only measure the fidelity of the graph state, with Pauli twirling, the graph state and T-doped graph state process fidelities are equal up to first order in infidelity<sup>5</sup> (We note that in this submission, we implement twirling on the CZ gates individually instead of in layers, which preserves the unstructured scheduling of the gates). We also make use of the ability to toggle between the undoped and doped circuits with rotational Z gates, which are implemented on IBM's hardware as noiseless framechanges<sup>6</sup>.

An additional argument for verification relies on analyzing the rate of logical faults (see Section C `Verifiability with Logical Fault Rate` in the [circuit instance description](https://github.com/quantum-advantage-tracker/quantum-advantage-tracker.github.io/tree/main/data/classically-verifiable-problems/circuit-models/doped_random_graph_sampling)). As the inserted T gates commute with the spacetime Pauli checks and are implemented noiselessly, we expect the post-selection or acceptance rate to be equal in the undoped and doped cases.

T gates therefore can only change the probability that a previously harmless fault (commutes with the stabilizers) becomes harmful and vice versa. The difference in fidelity between the graph and T-doped graph states can be expressed as

$$ F_{graph} - F_{doped} = P_{graph}(harmless\ |\ accept) - P_{doped}(harmless\ |\ accept) \leq max(P(harmless\ |\ accept))$$

where we numerically show that $P(harmless\ |\ accept)$ is small. So, the graph state fidelity closely verifies a lower bound for the doped graph state fidelity.

<p align="center">
<img width="3680" height="1050" alt="image" src="https://github.com/user-attachments/assets/7d0666d7-6c3d-4245-b0e6-d5cf7686f670" />

<em>Figure 1. Stochastic Pauli fault simulations for the 70 X 70 circuit. The model includes global depolarizing noise, decoherence error (200 μs), and readout error. Left: The probability of acceptance decreases exponentially with error rate. Middle: For accepted shots, the probability of no fault occuring is exponentially suppressed with increasing noise, while the probability of harmless and harmful faults occurring increases. Right: Only a small systematic difference exists between the fidelities of the graph (Clifford) and doped graph states across all error rates.</em>

</p>

As tensor network contraction costs scale with the entanglement (Schmidt rank) of the graph state, and extended Clifford simulators scale with the non-stabilizerness (stabilizer rank), we claim that sampling is adversarial for exact or naïve classical simulations.

## Experimental Details and Results

We prepare a T-doped random graph state on a 70 X 70 circuit on IBM Boston. We equip the circuit with 27 ancilla qubits and perform spacetime error detection.

<p align="center">
  <img width="468" height="322" alt="Image" src="https://github.com/user-attachments/assets/52ca7570-b809-49aa-b528-40ff33ba2311" />

<em>Figure 2. Physical layout for the 70 X 70 circuit (with 27 ancilla qubits) on IBM Boston, a Heron R3 device with heavy-hex architecture. The logical qubits (green) are arranged on a 1D chain with dangling ancilla qubits (red).</em>

</p>

Properties:

- **Graph state fidelity: 0.32** (bounded above 0.044 with 95% confidence)
- Data qubits: [45, 37, 25, 24, 23, 22, 21, 36, 41, 42, 43, 56, 63, 62, 61, 76, 81, 82, 83, 96, 103, 104, 105, 106, 107, 97, 87, 88, 89, 78, 69, 70, 71, 58, 51, 50, 49, 38, 29, 30, 31, 32, 33, 39, 53, 54, 55, 59, 75, 74, 73, 79, 93, 92, 91, 98, 111, 110, 109, 118, 129, 128, 127, 126, 125, 124, 123, 136, 143, 144]
- Ancillas: [44, 26, 16, 40, 64, 60, 80, 84, 102, 117, 108, 86, 90, 68, 72, 52, 48, 28, 18, 34, 94, 112, 130, 137, 122, 142, 145]
- Lowest known Schmidt rank (graph state): $2^{30}$
- Lowest known Stabilizer rank: $2^{185.5}$
- Maximum number of nontrivial T gates: 468

To verify the fidelity of the graph state, we employ direct fidelity estimation and measure the expectation values of 80 random stabilizers.

<p align="center">
  <img width="625" height="456" alt="image" src="https://github.com/user-attachments/assets/db6686f3-84f2-43b2-9758-6ab072339618" />
  
  <img width="617" height="456" alt="image" src="https://github.com/user-attachments/assets/f39d83d9-98f1-481a-94e4-21843610aec4" />

<em>Figure 3. Fidelity for the 70 X 70 circuit on IBM Boston when doping with 0, 468 S, and 468 Z gates. We utilize direct fidelity estimation with 80 randomly drawn Paulis. Fidelities are all consistent within one standard deviation, even after readout error mitigation. </em>

</p>

<p align="center">
  <img width="699" height="472" alt="image" src="https://github.com/user-attachments/assets/b5eaf9c1-8353-4504-ba00-e63dd58da873" />

<em>Figure 4. Post-selection rate for the 70 X 70 circuit on IBM Boston shows no dependence on S or Z doping count across the measured values. </em>

</p>

We do not directly measure the fidelity of the T-doped graph state. This is not efficient in general and, for direct fidelity estimation, would require exponentially many samples. As stated above, we instead infer the T-doped graph state fidelity to be equal to the graph state fidelity under the Pauli Twirling assumption and by analyzing logical faults.

## Classical Runtime Estimation

We estimate the runtime of two classical approaches, tensor networks (matrix product state) and extended stabilizer (Clifford + T) simulations.

For tensor networks, the entanglement is the most relevant challenge. To accurately represent a state, a bond dimension equivalent to the Schmidt rank is necessary. The contraction time is estimated for the 70 qubit circuit with increasing depth (max depth 24), with an estimated contraction time of $10^{25}$ seconds.

<p align="center">
  <img width="590" height="455" alt="image" src="https://github.com/user-attachments/assets/619e1f7d-77c1-4477-bddc-856a6706ec65" />

<em>Figure 5. Quimb matrix product state (MPS) contraction times for the 70 X 70 circuit with increasing depth (maximum depth 24). Linear extrapolation from a logarithmic plot of the data (R<sup>2</sup> > 0.98) yields a predicted contraction time of 10<sup>25</sup> seconds.</em>

</p>

For extended stabilizer simulations, the complexity scales with the non-stabilizerness of the state. We quantify this with the stabilizer rank, which depends exponentially on the size of the state. Simulations track the time to calculate one bitstring probability from 70 X 70 circuits with an increasing T-count (maximum 85). The estimated runtime is $10^{42}$ seconds.

<p align="center">
  <img width="570" height="476" alt="image" src="https://github.com/user-attachments/assets/f84b0143-1b2c-43c7-b9f4-dcfa4b8f0677" />

<em>Figure 6. Extended stabilizer simulation timing in quizx for 70 X 70 circuits with various T-counts (maximum 85). Linear extrapolation from a logarithmic plot of the data (R<sup>2</sup> > 0.98) yields a predicted time of 10<sup>42</sup> seconds to find the probability of one bitstring.</em>

</p>

## Classical Verification Task

The challenge for classical algorithms is to demonstrate strong simulation, either with the same fidelity in shorter time or with a higher fidelity in the same time as the quantum experiment. We propose three methods to verify that classical methods can adequately simulate random graph sampling:

1. Given that the T-doped graph state uses a maximum of $O(N^2)$ non-Clifford gates, it approximates a k-design<sup>7</sup> and is close to Haar-random. Therefore, the linear cross entropy benchmarking (XEB) score is a reasonable proxy for fidelity, given that the noise is weak and uncorrelated. We provide quantum samples for the 468 T-count instance below, for which one could show that a classical simulation produces an XEB score significantly larger than 0.044.
2. Alternatively, as in Zhao et. al's work on simulating random circuit sampling<sup>8</sup>, it also suffices to demonstrate that uncorrelated samples from the T-doped graph state can be produced with XEB score significantly larger than 0.044.
3. For classical simulations (e.g. tensor networks) for which it is equally difficult to represent the graph state and T-doped graph state, and for which conversion for the two can be done with small error, it would suffice to simulate the graph state, perform direct fidelity estimation, and demonstrate fidelity above 0.32.

Quantum Samples in the Z-basis (T-count of 468, 2051 bitstrings after post-selection):

- [samples_nq70_depth70_checks27_doped.npy.zip](https://github.com/user-attachments/files/29860637/samples_nq70_depth70_checks27_doped.npy.zip)
- (Bitstrings use Qiskit's little-endian notation)

## References

A detailed writeup can be found in the [circuit instance description](https://github.com/quantum-advantage-tracker/quantum-advantage-tracker.github.io/tree/main/data/classically-verifiable-problems/circuit-models/doped_random_graph_sampling).

1. [Ghosh, Hangleiter, Helsen. Random Regular Graph States are Complex at Almost Any Depth, PRX 2025](https://journals.aps.org/prxquantum/abstract/10.1103/52xz-3hpc)
2. [Arute et al. Quantum supremacy using a programmable superconducting processor, Nature 2019](https://www.nature.com/articles/s41586-019-1666-5)
3. [Martiel, Javadi-Abhari. Low-overhead error detection with spacetime codes, arXiv:2504.15725](https://arxiv.org/abs/2504.15725)
4. [Flammia, Liu. Direct Fidelity Estimation from Few Pauli measurements, PRA 2011](https://link.aps.org/doi/10.1103/PhysRevLett.106.230501)
5. [Merkel et al. When Clifford benchmarks are sufficient; estimating application performance with scalable proxy ](https://arxiv.org/pdf/2503.05943)[arXiv:2503.05943 ](https://arxiv.org/pdf/2503.05943)
6. [McKay et al. Efficient Z Gates for Quantum Computing, PRA 2018](https://link.aps.org/doi/10.1103/PhysRevA.96.022330)
7. [Leone et al. Non-Clifford Cost of Random Unitaries, arXiv:2505.10110](https://arxiv.org/pdf/2505.10110)
8. [Zhao et al. Leapfrogging Sycamore: Harnessing 1432 GPUs for 7× Faster Quantum Random Circuit Sampling, arXiv:2406.18889](https://arxiv.org/pdf/2406.18889)

### Quantum runtime (seconds)

17900

### Classical runtime (seconds)

10^25 (Quimb Matrix Product State)

10^42 (quizx Clifford + T)

### Compute resources (quantum)

IBM Heron R3

### Compute resources (classical)

Apple M1 Max 10 cores with 32 GB memory (Quimb), Intel Xeon Gold 6258R 28 cores with 1 TB memory (quizx Clifford + T)

### Notes

The reported classical runtimes are extrapolated.

### Authors

Ali Javadi-Abhari, Simon Martiel, Jay-U Chung, Alireza Seif, Ian Hincks, Jay Gambetta, Soumik Ghosh, Bill Fefferman

### Institutions

IBM, UChicago

**Comment by ajavadia** [comment]
@n-e-sherman This is very similar to the [previous post](https://github.com/quantum-advantage-tracker/quantum-advantage-tracker.github.io/issues/151) that you already approved. However this one is strictly harder classically, because now T gates can appear in the bulk of the Clifford skeleton as well, not just at the end.
Thanks for your time, let us know if you have any questions.

**Comment by n-e-sherman** [comment]
Hi @ajavadia and @jchung2020, thank you for the submission. Given the similarity to the previous submission, I have no additional questions, and I'm happy to accept the submission. I'll wait for @Roland-Farrell to review and provide any comments before formally accepting.

**Comment by Roland-Farrell** [comment]
Hi @ajavadia 
I just have one question about the submission:
- For the classical hardness you report the MPS contraction time in Quimb. What is the contraction time? Are you evaluating the circuit at a finite bond dimension, or trying to compute it exactly? 
Thanks

**Comment by jchung2020** [comment]
@Roland-Farrell 

Yes, we use Quimb to build an MPS (max bond dimension 2^12) of our 70 X 70 circuit at various depths from 2-24, and track the contraction time (when run on a laptop). This is what's shown in Figure 5.

**Comment by Roland-Farrell** [comment]
Thank you.
Have you looked at the fidelity obtained from the MPS if you instead cap the bond dimension to e.g. 600?

**Comment by jchung2020** [comment]
We have tried something similar. Recently we looked at fixing the max bond dimension and simulating the doped Clifford circuit with increasing depth until the truncation error exceeds than 1 - 0.32. The motivation was that Haar random circuits should be much easier to represent with an MPS because the singular value spectrum drops off exponentially. Indeed, we do see that the bond dimension does not scale as adversely for Haar random circuits, though it does for the doped Clifford circuit.

<img width="1152" height="720" alt="Image" src="https://github.com/user-attachments/assets/01605515-1c40-4d63-980d-64b45d01034a" />

The next plot shows the singular value spectrum for an MPS for a smaller 14 X 14 circuits, which is flat for Cliffords, exponentially decaying for Haar, and still exponentially decaying - but substantially less - for the doped Clifford circuits.

<img width="1728" height="720" alt="Image" src="https://github.com/user-attachments/assets/8ab3358d-a8d3-43fa-9511-f3684ea76a45" />

**Comment by ajavadia** [comment]
@Roland-Farrell Hi Roland, just wanted to see if you are good with this now?

**Comment by Roland-Farrell** [comment]
Thanks, so is the takeaway that truncating the bond dimension really doesn't buy you all that much?
For example, you see that you can reach a maximum depth ~ 25 before the bond dimension becomes prohibitively large, which is also the same depth you could reach before the contraction time became prohibitively long.

**Comment by jchung2020** [comment]
Yes, our conclusion is that truncation for MPS will not work well for our doped Clifford circuit because the singular value spectrum seems to be relatively flat. Not as flat as Clifford, but not as much of a decay as Haar-random, which means that we would need large bond dimensions to get close to our experimental fidelities.

**Comment by smartiel** [comment]
Hi Roland,
Here is a chat that demonstrates these spectra differences in the case of pure Clifford, doped Clifford and Haar random circuits.

<img width="642" height="282" alt="Image" src="https://github.com/user-attachments/assets/f2fee6c9-aed0-4755-a866-0b927ef58d2a" />

**Comment by Roland-Farrell** [comment]
Thanks, looks great!
I verified this submission.

**Comment by Nesoxy1** [comment]
Is there interest in a commit-reveal layer that proves the assertion threshold was fixed before the outcome was known, for forward-looking submissions? I have a working prototype.
